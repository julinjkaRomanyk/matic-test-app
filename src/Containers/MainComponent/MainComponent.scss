@import "utils/_variables";

.prog {
    width: 100%;
    height: 15px;
    background: $SecondOrange;
}

.MainComponent-Wrapper {
    display: flex;
    flex-direction: column;
    align-items: center;
}

.MainComponent {
    width: 100%;
    background: $BackgroundColor;
    margin-top: 84px;
}

@keyframes fadeIn {
    0% {
        display: none;
    }
    100% {
        display: flex;
    }
}

@keyframes fadeOut {
    0% {
        display: flex;
    }
    100% {
        display: none;
    }
}
