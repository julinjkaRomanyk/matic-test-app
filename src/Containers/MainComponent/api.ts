export type Question = {
  title: string;
  fields: Field[];
  questionNumber: number;
};

export type Field = {
  id: FieldId;
  label: string;
  type: string;
  validation?: string;
};

export enum FieldId {
  firstName = "firstName",
  lastName = "lastName",
  phoneNumber = "phoneNumber",
  stateName = "stateName"
}

export const questionArray: Question[] = [
    {
      "title": "What is Your Name?",
      questionNumber: 1,
      "fields": [
        {
          "id": FieldId.firstName,
          "label": "First Name",
          "type": "input",
          "validation": ""
        },
        {
          "id": FieldId.lastName,
          "label": "Last Name",
          "type": "input",
          "validation": ""
        }
      ]
    },
    {
      "title": "How Can We Reach Out to You?",
      questionNumber: 2,
      "fields": [
        {
          "id": FieldId.phoneNumber,
          "label": "Phone Number",
          "type": "input",
          "validation": "(999) 999-9999"
        }
      ]
    },
    {
      "title": "Where is Your House Located In?",
      questionNumber: 3,
      "fields": [
        {
          "id": FieldId.stateName,
          "label": "State",
          "type": "select"
        }
      ]
    }
  ]
  