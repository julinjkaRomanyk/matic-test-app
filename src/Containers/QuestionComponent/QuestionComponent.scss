.QuestionComponent-Wrapper {
    display: flex;
    flex-direction: column;
    align-items: center;

    @media all and (max-width: 736px) {
        padding: 10px;
    }
}

.QuestionComponent-QuestionTitle {
    font-size: 48px;

    @media all and (max-width: 736px) {
        font-size: 38px;
    }
}

.QuestionComponent-Input-Wrapper {
    display: flex;
    flex-direction: column;
    height: 20vh;
    justify-content: space-between;
    align-items: center;
    margin-top: 48px;
    margin-bottom: 24px;

    @media all and (max-width: 736px) {
        height: 40vh;
        width: unset;

        .Input-Wrapper {
            padding-bottom: 25px;
        }
    }
}

.QuestionComponent-Inputs-Wrapper {
    display: flex;
    justify-content: space-evenly;
    width: 75%;

    @media all and (max-width: 736px) {
        flex-direction: column;
        width: 100%;
    }
}
