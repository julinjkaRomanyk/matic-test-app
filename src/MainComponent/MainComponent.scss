.prog {
    width: 100%;
    height: 15px;
    background: #ffbe1e;
}

.MainComponent-Wrapper {
    display: flex;
    flex-direction: column;
    align-items: center;
}

.MainComponent {
    width: 100%;
    background: #fcfaf9;
    margin-top: 84px;
}

@keyframes fadeIn {
    0% {
        display: none;
    }
    100% {
        display: flex;
    }
}

@keyframes fadeOut {
    0% {
        display: flex;
    }
    100% {
        display: none;
    }
}
