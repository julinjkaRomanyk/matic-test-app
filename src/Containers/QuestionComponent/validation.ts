import { Reducer } from "react";
import { FieldId } from "Containers/MainComponent/api";

export const initialValidationErrors = {
  [FieldId.firstName]: "",
  [FieldId.lastName]: "",
  [FieldId.phoneNumber]: "",
  [FieldId.stateName]: ""
};

export type ValidationErrors = {
  firstName: string,
  lastName: string,
  phoneNumber: string,
  stateName: string
}

export const validationErrorsReducer: Reducer<ValidationErrors, Partial<ValidationErrors>> = (state, newState) => ({
  ...state,
  ...newState
});


export const initializeValidators = (setValidationErrors: React.Dispatch<Partial<ValidationErrors>>) => ({
  firstNameValid(name: string) {
    const trimmedName = name.trim();
    if (!trimmedName) {
      setValidationErrors({
        firstName: "This field is required"
      });

      return false;
    }

    if (trimmedName.length < 3) {
      setValidationErrors({
        firstName: "Should contain at least 3 characters"
      });

      return false;
    }

    if (trimmedName.length > 128) {
      setValidationErrors({
        firstName: "Should not be greater than 128 characters"
      });

      return false;
    }

    setValidationErrors({
      firstName: ""
    })

    return true;
  },

  lastNameValid(name: string) {
    const trimmedName = name.trim();
    if (!trimmedName) {
      setValidationErrors({
        lastName: "This field is required"
      });

      return false;
    }

    if (trimmedName.length < 3) {
      setValidationErrors({
        lastName: "Should contain at least 3 characters"
      });

      return false;
    }

    if (trimmedName.length > 128) {
      setValidationErrors({
        lastName: "Should not be greater than 128 characters"
      });

      return false;
    }

    setValidationErrors({
      lastName: ""
    });

    return true;
  },

  phoneNumberValid(number: string) {
    const trimmedPhoneNumber = number.trim();
    if (!trimmedPhoneNumber) {
      setValidationErrors({
        phoneNumber: "This field is required"
      });

      return false;
    }

    if (!number.match(/^\(?([0-9]{3})\)?[-. ]?([0-9]{3})[-. ]?([0-9]{4})$/)) {
      setValidationErrors({
        phoneNumber: "Seems like this number is not valid"
      });

      return false;
    }

    setValidationErrors({
      phoneNumber: ""
    });

    return true;
  },

  stateNameValid(state: string) {
    if (!state) {
      setValidationErrors({
        stateName: "This field is required"
      });

      return false;
    }

    setValidationErrors({
      stateName: ""
    });

    return true;
  }
});