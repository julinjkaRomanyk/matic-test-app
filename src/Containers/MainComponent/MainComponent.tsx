import React, { useState } from "react";

import { questionArray } from "./api";
import { QuestionComponent } from "Containers/QuestionComponent";
import { Fade } from "Components/AnimationWrapper";
import { ProgressBarComponent } from "Components/ProgressBar";
import { StartEndComponent } from "Components/StartEndComponent";
import HomeImage from "images/HomeImage.png";
import ThankYouImage from "images/ThankYouImage.png";
import { welcomeMessage, getEndComponentQuestion } from "utils/index";

import { InfoStore } from "utils/types";
import "./MainComponent.scss"

export const initialStoreInfo: InfoStore = {
    firstName: "",
    lastName: "",
    phoneNumber: "",
    stateName: "",
}

export const MainComponent: React.FC = () => {
    const [infoStore, setInfoStore] = useState<InfoStore>(initialStoreInfo);
    const [questionCounter, setQuestionCounter] = useState<number>(0);

    const handleStoreChange = (changes: any) => {
        setInfoStore({ ...infoStore, ...changes });
    }

    const handleQuestionNumberChange = (counter: number) => {
        setQuestionCounter(counter + 1);
    }

    const shouldRenderStartComponent = questionCounter === 0;
    const shouldRenderEndComponent = questionCounter === questionArray.length + 1 ;

    const renderQuestionComponent = () => {
        return questionArray.map(item => {

            return (
                <Fade show={questionCounter === item.questionNumber} key={item.questionNumber}>
                    <QuestionComponent
                        questionTitle={item.title}
                        questionNumber={item.questionNumber}
                        fields={item.fields}
                        infoStore={infoStore}
                        onValueChange={handleStoreChange}
                        onButtonClick={handleQuestionNumberChange}
                        buttonLabel={item.questionNumber === questionArray.length ? "Finish" : "Next"}
                    />
                </Fade>
            );
        })
    }

    const renderStartEndDateComponent = () => {
        return (
            <>
                {shouldRenderStartComponent && <StartEndComponent message={welcomeMessage} buttonLabel="Start" onButtonClick={() => setQuestionCounter(1)} image={HomeImage} />}
                {shouldRenderEndComponent && <StartEndComponent message={getEndComponentQuestion(infoStore.firstName, infoStore.lastName, infoStore.phoneNumber)} image={ThankYouImage} />}
            </>
        );

    }

    return (
        <div className="MainComponent-Wrapper" >
            { (!shouldRenderStartComponent && !shouldRenderEndComponent) && <ProgressBarComponent progress={{ current: questionCounter, total: questionArray.length }} /> }
            {renderStartEndDateComponent()}
            <div className="MainComponent">
                {renderQuestionComponent()}
            </div>

        </div>
    );
}