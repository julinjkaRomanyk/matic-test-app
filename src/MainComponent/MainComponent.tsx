import React, { useState } from "react";

import { questionArray } from "./api";
import { QuestionComponent } from "./../QuestionComponent";
import { Fade } from "./../QuestionComponent/AnimationWrapper";
import { ProgressBarComponent } from "./../ProgressBar";
import { StartEndComponent } from "./../StartEndComponent";
import HomeImage from "./../images/HomeImage.png";
import ThankYouImage from "./../images/ThankYouImage.png";
import { welcomeMessage, getEndComponentQuestion } from "./../utils/index";

import { InfoStore } from "./../utils/types";
import "./MainComponent.scss"

export const initialStoreInfo: InfoStore = {
    firstName: "",
    lastName: "",
    phoneNumber: "",
    stateName: "",
}

export const MainComponent: React.FC = () => {
    const [infoStore, setInfoStore] = useState<InfoStore>(initialStoreInfo);
    const [questionCounter, setQuestionCounter] = useState<number>(0);

    const handleStoreChange = (changes: any) => {
        setInfoStore({ ...infoStore, ...changes });
    }

    const handleQuestionNumberChange = (counter: number) => {
        setQuestionCounter(counter + 1);
    }

    const renderQuestionComponent = () => {
        return questionArray.map(item => {

            return (
                <Fade show={questionCounter === item.questionNumber} key={item.questionNumber}>
                    <QuestionComponent
                        questionTitle={item.title}
                        questionNumber={item.questionNumber}
                        fields={item.fields}
                        infoStore={infoStore}
                        onValueChange={handleStoreChange}
                        onButtonClick={handleQuestionNumberChange}
                        buttonLabel={item.questionNumber === questionArray.length ? "Finish" : "Next"}
                    />
                </Fade>
            );
        })
    }

    const renderStartEndDateComponent = () => {
        return (
            <>
                {questionCounter === 0 && <StartEndComponent message={welcomeMessage} buttonLabel="Start" onButtonClick={() => setQuestionCounter(1)} image={HomeImage} />}
                {questionCounter === questionArray.length + 1 && <StartEndComponent message={getEndComponentQuestion(infoStore.firstName, infoStore.lastName, infoStore.phoneNumber)} image={ThankYouImage} />}
            </>
        );

    }

    return (
        <div className="MainComponent-Wrapper" >
            { questionCounter !== 0  && <ProgressBarComponent progress={{ current: questionCounter, total: questionArray.length }} /> }
            {renderStartEndDateComponent()}
            <div className="MainComponent">
                {renderQuestionComponent()}
            </div>

        </div>
    );
}