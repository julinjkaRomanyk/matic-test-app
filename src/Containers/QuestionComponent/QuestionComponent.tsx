import React, {useRef, useReducer} from "react";

import { Button } from "Components/Button";
import { Input } from "Components/Input";
import { Select } from "Components/SearchSelect";

import "./QuestionComponent.scss"
import { Field, FieldId } from "Containers/MainComponent/api";
import { InfoStore } from "utils/types";
import { statesList } from "utils/index";
import { validationErrorsReducer, initialValidationErrors, initializeValidators } from "./validation";

export type Props = {
    questionTitle: string;
    questionNumber: number;
    buttonLabel?: string;
    fields: Field[];
    infoStore: InfoStore;

    onValueChange(changes: any): void;
    onButtonClick(number: number): void;

}

export const QuestionComponent: React.FC<Props> = ({ questionTitle, buttonLabel, onButtonClick, onValueChange, questionNumber,  fields, infoStore }) => {
    const [validationErrors, setValidationErrors] = useReducer(validationErrorsReducer, initialValidationErrors);
    const { current: validators } = useRef(initializeValidators(setValidationErrors));
 

        const handleButtonClick = () => {
            // All switch cases can be refactored in favor of a small method that could automatically take RegExp
            //  from the api and just test field with proper regex string and return default "Invalid value" error message
            // So later I'd like to explain my choice of creating separate error form

            const isFormValid = fields.map(field => {
                switch (field.id) {
                    case FieldId.firstName: return validators.firstNameValid(infoStore[field.id]);
                    case FieldId.lastName: return validators.lastNameValid(infoStore[field.id]);
                    case FieldId.phoneNumber: return validators.phoneNumberValid(infoStore[field.id]);
                    case FieldId.stateName: return validators.stateNameValid(infoStore[field.id]);
    
                    default: return true;
                }
            });
    
            const validFields = isFormValid.every(field => field)
    
            if(validFields) {
                onButtonClick(questionNumber)
                }

        }

        const handleValueChange = (value: string, fieldId: FieldId  ) => {
            setValidationErrors({...validationErrors, [fieldId] : "" })
            onValueChange({ [fieldId] : value })
        }

        const selectProps = {
            placeholder: "Select",
            options: statesList,
          };

        const getQuestionFields = (fields : Field[]) => {
            return (
                <>
                    <div className="QuestionComponent-Inputs-Wrapper">
                        {
                            fields.map((field: Field) => {
                                switch (field.type) {
                                    case "input": return <Input key={questionNumber + field.id} error={validationErrors[field.id]} label={field.label} onChange={(event) => handleValueChange(event, field.id )} mask={field.mask} />;           
                                    case "select": return <Select key={questionNumber + field.id} error={validationErrors[field.id]} selectProps={{...selectProps, onChange: (state: any) => handleValueChange(state.value, field.id) }} label={field.label} />;
                                    default: return <Input  key={questionNumber + field.id} onChange={ () => {} } />;
                                }
                            })
                        }
                    </div>
                    <Button label={buttonLabel} onClick={handleButtonClick} data-testid="button-callback" />
                </>
            )
         
        };

    return (
        <div className="QuestionComponent-Wrapper">
            <div className="QuestionComponent-QuestionTitle">
                {questionTitle}
            </div>
            <div className="QuestionComponent-Input-Wrapper">
            {getQuestionFields(fields)}
            </div>
        </div>
    )
};