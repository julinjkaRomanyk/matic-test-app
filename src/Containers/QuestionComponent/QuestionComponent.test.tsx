import React from "react";
import { getTestName } from "utils/test-utils";
import { render, fireEvent, act } from "@testing-library/react";

import  { FieldId } from "Containers/MainComponent/api";

import { QuestionComponent } from "./QuestionComponent";

export const validStoreInfo = {
    firstName: "Julia",
    lastName: "Romanyk",
    phoneNumber: "0987898909",
    stateName: "",
}

export const invalidStoreInfo = {
    firstName: "",
    lastName: "",
    phoneNumber: "",
    stateName: "",
}


const mockProps = {
    questionTitle: "What is Your Name?",
    questionNumber: 1,
    fields: [
        {
          "id": FieldId.firstName,
          "label": "First Name",
          "type": "input",
          "validation": "^.{3,128}$"
        },
        {
          "id": FieldId.lastName,
          "label": "Last Name",
          "type": "input",
          "validation": "^.{3,128}$"
        }
      ],
    infoStore: validStoreInfo,

    onValueChange: jest.fn(),
    onButtonClick: jest.fn(),
}

describe(getTestName(__dirname), () => {
    test("Match snapshot", () => {
      const { container } = render(<QuestionComponent {...mockProps}/>);
  
      expect(container.firstChild).toMatchSnapshot();
    });

    test(`should trigger onButtonClick callback, when button is pressed and inputs are filled with valid info`, async () => {
        const { getByTestId } = render(<QuestionComponent {...mockProps}/>);

        const saveButton = getByTestId("button");

          fireEvent.click(saveButton);

        expect(mockProps.onButtonClick).toHaveBeenCalled();
      });
  });
  