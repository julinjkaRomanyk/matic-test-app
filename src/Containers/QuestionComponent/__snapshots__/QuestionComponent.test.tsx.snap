// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionComponent Match snapshot 1`] = `
<div
  class="QuestionComponent-Wrapper"
>
  <div
    class="QuestionComponent-QuestionTitle"
  >
    What is Your Name?
  </div>
  <div
    class="QuestionComponent-Input-Wrapper"
  >
    <div
      class="QuestionComponent-Inputs-Wrapper"
    >
      <div
        class="Input-Wrapper "
      >
        <span
          class="Input-Label"
          data-testid="Input-Label"
        >
           
          First Name
           
        </span>
        <input
          data-testid="Input"
          value=""
        />
      </div>
      <div
        class="Input-Wrapper "
      >
        <span
          class="Input-Label"
          data-testid="Input-Label"
        >
           
          Last Name
           
        </span>
        <input
          data-testid="Input"
          value=""
        />
      </div>
    </div>
    <button
      class="Button"
      data-testid="button"
      id="button"
    />
  </div>
</div>
`;
